Tarquin Bennett

# Question 1

### Your Turn 1

In [9]:
5*3+(20*9)

195

In [6]:
(8*90+12)/7

104.57142857142857

In [7]:
23*45-(123*456)

-55053

In [10]:
12345*54321

670592745

### Your Turn 2

In [2]:
import nltk

In [3]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

In [5]:
from nltk.book import *

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [4]:
text2.concordance("affection")

Displaying 25 of 79 matches:
, however , and , as a mark of his affection for the three girls , he left them
t . It was very well known that no affection was ever supposed to exist between
deration of politeness or maternal affection on the side of the former , the tw
d the suspicion -- the hope of his affection for me may warrant , without impru
hich forbade the indulgence of his affection . She knew that his mother neither
rd she gave one with still greater affection . Though her late conversation wit
 can never hope to feel or inspire affection again , and if her home be uncomfo
m of the sense , elegance , mutual affection , and domestic comfort of the fami
, and which recommended him to her affection beyond every thing else . His soci
ween the parties might forward the affection of Mr . Willoughby , an equally st
 the most pointed assurance of her affection . Elinor could not be surprised at
he natural consequence of a strong affection in a young and ardent mind . This 
 opinion . 

In [5]:
text3.concordance("lived")

Displaying 25 of 38 matches:
ay when they were created . And Adam lived an hundred and thirty years , and be
ughters : And all the days that Adam lived were nine hundred and thirty yea and
nd thirty yea and he died . And Seth lived an hundred and five years , and bega
ve years , and begat Enos : And Seth lived after he begat Enos eight hundred an
welve years : and he died . And Enos lived ninety years , and begat Cainan : An
 years , and begat Cainan : And Enos lived after he begat Cainan eight hundred 
ive years : and he died . And Cainan lived seventy years and begat Mahalaleel :
rs and begat Mahalaleel : And Cainan lived after he begat Mahalaleel eight hund
years : and he died . And Mahalaleel lived sixty and five years , and begat Jar
s , and begat Jared : And Mahalaleel lived after he begat Jared eight hundred a
and five yea and he died . And Jared lived an hundred sixty and two years , and
o years , and he begat Eno And Jared lived after he begat Enoch eight hundred y
 and two ye

In [6]:
text4.concordance("terror")

Displaying 8 of 8 matches:
menaces , by fraud or violence , by terror , intrigue , or venality , the Gove
ameless , unreasoning , unjustified terror which paralyzes needed efforts to c
ublic seemed frozen by a fatalistic terror , we proved that this is not true .
 to alter that uncertain balance of terror that stays the hand of mankind ' s 
eans freeing all Americans from the terror of runaway living costs . All must 
still . They fuel the fanaticism of terror . And they torment the lives of mil
d maintain a strong defense against terror and destruction . Our children will
k to advance their aims by inducing terror and slaughtering innocents , we say


In [7]:
text5.concordance("ur")

Displaying 21 of 21 matches:
k up PART no i dont want daughters !! ur annoying . ACTION Now Playing - Cradl
ooo U92 !!!!!! . ACTION U1370 watches ur nad with a stick . ca u U23 ewwww lol
er lip . Meep . ACTION is resisting . ur female right lol U115 beeeeehave Reme
 geeshh ... two kids fighting ! whats ur major i probably sucks in summer too 
II Men scorpions rock ... lol what is ur job me too U11 hehe went to manhattan
e no one i like to say if u have done ur time then that is that U42 Ok U37 , i
hat a ride JOIN ty ty cheers babes .. ur the first to follow up with the pseud
 charger \ty LoL yeah ;-) well i hope ur doing ok .. i 'm dojn fine babe . . g
 gonna rock up soojn and rob u of all ur ' candy ' :) as long as you are happy
genital warts ? LoL moped U28 ?.. ohh ur a real man i had a moped once yup heh
b-day is in 5 days =( PART LOL U35 no ur nawt yup your gay lmao U37 dang come 
 people talk to me anymore oh because ur gay . PART PART . ACTION = U58 . Man 
 om hi U73 heya U7 ! h 

In [8]:
text6.concordance("holy")

Displaying 25 of 26 matches:
[ angels sing ] Arthur , this is the Holy Grail . Look well , Arthur , for it 
rpose , Arthur ... the quest for the Holy Grail . [ boom ] [ singing stops ] L
 he can join us in our quest for the Holy Grail . FRENCH GUARD : Well , I ' ll
gy was required if the quest for the Holy Grail were to be brought to a succes
GALAHAD : You are the keepers of the Holy Grail ? ZOOT : The what ? GALAHAD : 
CELOT : No , we ' ve got to find the Holy Grail . Come on ! GALAHAD : Oh , let
uld be the sign that leads us to the Holy Grail ! Brave , brave Concorde , you
ung people in the joyful bond of the holy wedlock . Unfortunately , one of the
' ve not given up your quest for the Holy Grail ? MINSTREL : [ singing ] He is
 TIM : I do . [ zoosh ] You seek the Holy Grail ! ARTHUR : That is our quest .
, we ' re -- we ' re looking for the Holy Grail . Our quest is to find the Hol
oly Grail . Our quest is to find the Holy Grail . KNIGHTS : Yeah . Yes . It is
 TIM : Yes , I can help

This shows how diverse a single word can be and how it can be used in several different ways. 

### Your Turn 3

In [20]:
text2.similar("affection")

attention time regard mother love heart opinion sister wishes wife
arrival marianne kindness family it marriage sisters sake conduct mind


In [42]:
text4.similar("affection")

freedom that hope influence policy justice manufactures peace property
each efficiency civilization and to life me which transmitted order
but


In [28]:
text2.common_contexts(["affection","love"])

his_for her_for s_for his_was s_and and_to s_it her_and


In [43]:
text4.common_contexts(["affection","freedom"])

and_which and_to of_the


This shows that within two different text that the word "affection" can have two different meanings. The first text shows that the author's connotation is more towards love of another human, while text 4 was more about love of a country.

### Your Turn 4

In [44]:
text5.count('lol')

704

In [47]:
(text5.count('lol')/len(set(text5)))*100

11.60567095285196

This means that within the subset of chats they collected that 11.6% of the chat was lol. 

### Your Turn 5

In [52]:
lst1 = ["Hello", "World","!"]

In [53]:
lst2 = ["The","quick","brown","fox","jumps","over","the","lazy","dog"]

In [49]:
sorted(lst1)

['!', 'Hello', 'World']

In [51]:
len(set(lst1))

3

In [54]:
len(lst2)

9

In [59]:
lst2.count("the")

1

In [61]:
lst1+lst2

['Hello',
 'World',
 '!',
 'The',
 'quick',
 'brown',
 'fox',
 'jumps',
 'over',
 'the',
 'lazy',
 'dog']

This just shows that you can interact with list just like the text files.

### Your Turn 6

In [67]:
sent=["Python","is","a","pretty","great","language"]

In [68]:
sent

['Python', 'is', 'a', 'pretty', 'great', 'language']

In [69]:
sent[4] = "bad" 

In [70]:
sent

['Python', 'is', 'a', 'pretty', 'bad', 'language']

In [72]:
sent[1:4]

['is', 'a', 'pretty']

In [73]:
sent[1:4] = ['is', 'not', 'a']

In [74]:
sent

['Python', 'is', 'not', 'a', 'bad', 'language']

list are powerful tools when it comes to editing and displaying text

### Your Turn 7

In [81]:
fdist2 = FreqDist(text2)

In [82]:
print(fdist2)

<FreqDist with 6833 samples and 141576 outcomes>


In [83]:
fdist2.most_common(50)

[(',', 9397),
 ('to', 4063),
 ('.', 3975),
 ('the', 3861),
 ('of', 3565),
 ('and', 3350),
 ('her', 2436),
 ('a', 2043),
 ('I', 2004),
 ('in', 1904),
 ('was', 1846),
 ('it', 1568),
 ('"', 1506),
 (';', 1419),
 ('she', 1333),
 ('be', 1305),
 ('that', 1297),
 ('for', 1234),
 ('not', 1212),
 ('as', 1179),
 ('you', 1037),
 ('with', 971),
 ('had', 969),
 ('his', 941),
 ('he', 895),
 ("'", 883),
 ('have', 807),
 ('at', 806),
 ('by', 737),
 ('is', 728),
 ('."', 721),
 ('s', 700),
 ('Elinor', 684),
 ('on', 676),
 ('all', 642),
 ('him', 633),
 ('so', 617),
 ('but', 597),
 ('which', 592),
 ('could', 568),
 ('Marianne', 566),
 ('my', 551),
 ('Mrs', 530),
 ('from', 527),
 ('would', 507),
 ('very', 492),
 ('no', 488),
 ('their', 463),
 ('them', 462),
 ('--', 461)]

In [86]:
fdist2["Elinor"]

684

This list is the first 50 most common words in text2. This list does not really tell us much of the text other than there are two characters named Elinor and Marianne, and since they occur a lot in the text we can deduce that they could be the main characters.

# Question 2

### 2.4

In [162]:
a='leda'
b='deal'

n=len(a)
m=len(b)

D=[[0 for _ in range(m+1)] for _ in range(n+1)]

for i in range(1,n+1):
    D[i][0] = D[i-1][0]+1

for i in range(1,m+1):
    D[0][i] = D[0][i-1]+1

for i in range(1,n+1):
    for j in range(1,m+1):
        if a[i-1]==b[j-1]:
            D[i][j] = D[i-1][j-1]
        else:
            D[i][j] = min(D[i-1][j]+1,D[i-1][j-1]+1,D[i][j-1]+1)
#prints the grid
for i in D:
    print(i)
print(D[n][m])

[0, 1, 2, 3, 4]
[1, 1, 2, 3, 3]
[2, 2, 1, 2, 3]
[3, 2, 2, 2, 3]
[4, 3, 3, 2, 3]
3


The min edit distance is 3. This means that you need a combo of insertions, deletions, and substitutions which totals 3 to get from leda to deal.

### 2.5

cost of substitution is 2

In [160]:
a='drive'
b='brief'
n=len(a)
m=len(b)

D=[[0 for _ in range(m+1)] for _ in range(n+1)]

for i in range(1,n+1):
    D[i][0] = D[i-1][0]+1

for i in range(1,m+1):
    D[0][i] = D[0][i-1]+1

for i in range(1,n+1):
    for j in range(1,m+1):
        if a[i-1]==b[j-1]:
            D[i][j]=D[i-1][j-1]
        else:
            D[i][j] = min(D[i-1][j]+1,D[i-1][j-1]+1,D[i][j-1]+1)
#prints the grid
for i in D:
    print(i)
print(D[n][m])

[0, 1, 2, 3, 4, 5]
[1, 1, 2, 3, 4, 5]
[2, 2, 1, 2, 3, 4]
[3, 3, 2, 1, 2, 3]
[4, 4, 3, 2, 2, 3]
[5, 5, 4, 3, 2, 3]
3


In [159]:
a='drive'
b='drivers'
n=len(a)
m=len(b)

D=[[0 for _ in range(m+1)] for _ in range(n+1)]

for i in range(1,n+1):
    D[i][0] = D[i-1][0]+1

for i in range(1,m+1):
    D[0][i] = D[0][i-1]+1

for i in range(1,n+1):
    for j in range(1,m+1):
        if a[i-1]==b[j-1]:
            D[i][j]=D[i-1][j-1]
        else:
            D[i][j] = min(D[i-1][j]+1,D[i-1][j-1]+1,D[i][j-1]+1)
#prints the grid
for i in D:
    print(i)
print(D[n][m])

[0, 1, 2, 3, 4, 5, 6, 7]
[1, 0, 1, 2, 3, 4, 5, 6]
[2, 1, 0, 1, 2, 3, 4, 5]
[3, 2, 1, 0, 1, 2, 3, 4]
[4, 3, 2, 1, 0, 1, 2, 3]
[5, 4, 3, 2, 1, 0, 1, 2]
2


Since the edit distance between drive and drivers is 2, compared to drive and brief which is 3, that means drivers is closer to drive than brief is.

This means that we have to do less insertions, deletions, and/or substitutions to get from drive to drivers then to brief.